scripts/filter_expected_warnings.py

```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/env python
+"""Script to filter out old warnings from doxygen."""
+
+import sys
+import re
+
+
+class DoxygenWarning(object):
+    """Doxygen warning class."""
+
+    def __init__(self, firstline, filename, warning):
+        self.firstline = firstline
+        self.filename = filename
+        self.warning = warning
+        self.otherlines = []
+
+    def equals_ignoring_path_and_line_number(self, other):
+        """Return true if warnings have same filename and warning message."""
+        if self.filename != other.filename:
+            return False
+        if self.warning != other.warning:
+            return False
+        return self.otherlines == other.otherlines
+
+    def print_with_prefix(self, prefix):
+        print(prefix + self.firstline)
+        for line in self.otherlines:
+            print(prefix + '  ' + line)
+
+
+class WarningsList(object):
+    """List of Doxygen warnings."""
+
+    def __init__(self, all_lines):
+        """Create a list of the warnings in this file."""
+        self.warnings_list = []
+        current = None
+
+        warning_start_expr = re.compile(r'[^:]+/([^:/]+):\d+: (warning.*$)')
+
+        for line in all_lines:
+            if line.isspace():
+                continue
+
+            # Allow comments in the list of expected warnings.
+            if line.startswith('#'):
+                continue
+
+            matched = warning_start_expr.match(line)
+            if matched:
+                filename = matched.group(1)
+                warning = matched.group(2)
+                current = DoxygenWarning(line.strip(), filename, warning)
+                self.warnings_list.append(current)
+            elif line.startswith('  '):
+                current.otherlines.append(line.strip())
+            else:
+                # Assuming all warnings are of the form [path:line: warning:...]
+                # (and the warnings about too many nodes have been filtered out).
+                print('Error filtering warnings: Unexpected input format.')
+                print('  Input:' + line)
+
+    def contains(self, warning):
+        """Check if a similar warning is in this list."""
+        for other in self.warnings_list:
+            if warning.equals_ignoring_path_and_line_number(other):
+                return True
+        return False
+
+    def is_not_subset_of(self, superset, prefix):
+        """Check if this a subset of superset, and print anything missing."""
+        missing_element_found = False
+        for warning in self.warnings_list:
+            if not superset.contains(warning):
+                missing_element_found = True
+                warning.print_with_prefix(prefix)
+        return missing_element_found
+
+
+def ignore_too_many_nodes(all_lines):
+    """Filter out lines about graphs with too many nodes."""
```

Comment:
:question: Why aren't these just in the ignore list - please elaborate in the comment why we don't care about this warning

Reply:
Handling them in this way (rather than putting them in the ignore list) means that any future similar errors are also ignored.  This was done in the old scripts, and there was no explanation there as to why.  I can think of good reasons for doing this, but I'd prefer not to comment on it, because it was someone else's decision.

(For example, the graphs probably get too big to be useful if they have too many nodes.  If DOT_GRAPH_MAX_NODES is increased, it should be because the missing graphs in the documentation are found to be an issue - not because of the warning message.)